bundles/org.openhab.ui/web/src/assets/definitions/blockly/utils.js

```diff
@@ -13,31 +13,128 @@ export function addOSGiService (serviceName, serviceClass) {
+    'getZonedDateTime', [
+      '/* Try to detect the format based on its length */',
+      'function ' + Blockly.JavaScript.FUNCTION_NAME_PLACEHOLDER_ + '(datetime) {',
+      '  datetime = datetime.replace(\'T\', \' \')',
+      '  switch (datetime.length) {',
+      `    case 10: return ${zdt}.parse(datetime + ' 00:00:00+00:00', dtf.ofPattern('yyyy-MM-dd HH:mm:ssz'));`,
+      `    case 16: return ${zdt}.parse(datetime + ':00+00:00', dtf.ofPattern('yyyy-MM-dd HH:mm:ssz'));`,
+      `    case 19: return ${zdt}.parse(datetime + '+00:00', dtf.ofPattern('yyyy-MM-dd HH:mm:ssz'));`,
+      `    case 25: return ${zdt}.parse(datetime, dtf.ofPattern('yyyy-MM-dd HH:mm:ssz'));`,
+      `    case 23: return ${zdt}.parse(datetime + ' +00:00', dtf.ofPattern('yyyy-MM-dd HH:mm:ss.SSS z'));`,
+      `    case 26: return ${zdt}.parse(datetime + ' +00:00', dtf.ofPattern('yyyy-MM-dd HH:mm:ss.SSSSSS z'));`,
+      `    case 29: return ${zdt}.parse(datetime, dtf.ofPattern('yyyy-MM-dd HH:mm:ss.SSSSz'));`,
+      `    case 32: return ${zdt}.parse(datetime, dtf.ofPattern('yyyy-MM-dd HH:mm:ss.SSSSSSSz'));`,
+      `    case 28: return ${zdt}.parse(datetime.slice(0,26) + ':' + datetime.slice(26,28), dtf.ofPattern('yyyy-MM-dd HH:mm:ss.SSSSz'));`,
+      `    default: return ${zdt}.parse(datetime);`,
+      '  }',
+      '}'
+    ])
+
+  let createzdt = Blockly.JavaScript.provideFunction_(
+    'createZonedDateTime', [
+      'function ' + Blockly.JavaScript.FUNCTION_NAME_PLACEHOLDER_ + '(year, month, day, hour, minute, second, nano, offsetString, timezoneString) {',
+      '  stringToParse = \'\' + year;',
+      '  stringToParse += \'-\' + (\'0\' + month).slice(-2);',
+      '  stringToParse += \'-\' + (\'0\' + day).slice(-2);',
+      '  stringToParse += \'T\' + (\'0\' + hour).slice(-2);',
+      '  stringToParse += \':\' + (\'0\' + minute).slice(-2);',
+      '  stringToParse += \':\' + (\'0\' + second).slice(-2);',
+      '  stringToParse += \'.\' + nano + offsetString + \'[\' + timezoneString + \']\';',
+      `  return ${zdt}.parse(stringToParse, dtf.ISO_ZONED_DATE_TIME);`,
+      '}'
+    ])
+
+  return [dtf, zdt, getzdt, createzdt]
+}
+
+export function addGetZdtComponent () {
+  let getZdtComponent = Blockly.JavaScript.provideFunction_(
+    'getZdtComponent', [
+      'function ' + Blockly.JavaScript.FUNCTION_NAME_PLACEHOLDER_ + '(value) {',
+      '  return (typeof value == \'number\') ? value : value.getValue();',
+      '}'
+    ])
+
+  return getZdtComponent
+}
+
+export function addChrono () {
+  let chrono = Blockly.JavaScript.provideFunction_(
+    'chrono',
+    ['var ' + Blockly.JavaScript.FUNCTION_NAME_PLACEHOLDER_ + ' = Java.type("java.time.temporal.ChronoField");'])
+  return chrono
+}
+
+export function addDateComparisonSupport () {
+  let zdtCompare = Blockly.JavaScript.provideFunction_(
+    'zdtCompare', [
+      'function ' + Blockly.JavaScript.FUNCTION_NAME_PLACEHOLDER_ + '(zdt1, zdt2, compareOp, precision, compDate) {',
```

Comment:
Similarly, instead of injecting this huge utility function, since those compareOp/precision/compDate are fields, you could generate the code dynamically, using [temporary variables](https://developers.google.com/blockly/guides/create-custom-blocks/caching-arguments#temporary_variables) to store copies of  `zdt1` & `zdt2` with the "precision" and "compDate" logic, then generate the code that you need dynamically.
In other words this big switchs/cases utility function should have been implemented in the code generation for the blocks instead.

Reply:
I will add that to my todo list and see what I can do here.